<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Generate a Circular Cladogram with Heatmap for Taxa — generate_taxa_cladogram_single • MicrobiomeStat</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Generate a Circular Cladogram with Heatmap for Taxa — generate_taxa_cladogram_single"><meta name="description" content="This function generates a circular cladogram with an integrated heatmap for taxonomic data.
It visualizes the phylogenetic relationships between different taxa and their abundances or other
coefficients across different taxonomic levels using a tree-like structure (cladogram)."><meta property="og:description" content="This function generates a circular cladogram with an integrated heatmap for taxonomic data.
It visualizes the phylogenetic relationships between different taxa and their abundances or other
coefficients across different taxonomic levels using a tree-like structure (cladogram)."><meta property="og:image" content="https://cafferychen777.github.io/MicrobiomeStat/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MicrobiomeStat</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/cafferychen777/MicrobiomeStat"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Generate a Circular Cladogram with Heatmap for Taxa</h1>
      <small class="dont-index">Source: <a href="https://github.com/cafferychen777/MicrobiomeStat/blob/main/R/generate_taxa_cladogram_single.R" class="external-link"><code>R/generate_taxa_cladogram_single.R</code></a></small>
      <div class="d-none name"><code>generate_taxa_cladogram_single.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function generates a circular cladogram with an integrated heatmap for taxonomic data.
It visualizes the phylogenetic relationships between different taxa and their abundances or other
coefficients across different taxonomic levels using a tree-like structure (cladogram).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">generate_taxa_cladogram_single</span><span class="op">(</span></span>
<span>  <span class="va">data.obj</span>,</span>
<span>  <span class="va">test.list</span>,</span>
<span>  group.var <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">feature.level</span>,</span>
<span>  feature.mt.method <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  color.group.level <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  palette <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  pdf <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  pdf.width <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  pdf.height <span class="op">=</span> <span class="fl">10</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data-obj">data.obj<a class="anchor" aria-label="anchor" href="#arg-data-obj"></a></dt>
<dd><p>A list object in a format specific to MicrobiomeStat, which includes components like feature.tab, feature.ann, meta.dat, etc.</p></dd>


<dt id="arg-test-list">test.list<a class="anchor" aria-label="anchor" href="#arg-test-list"></a></dt>
<dd><p>A list of test results. If NULL, it will be generated using generate_taxa_test_single.</p></dd>


<dt id="arg-group-var">group.var<a class="anchor" aria-label="anchor" href="#arg-group-var"></a></dt>
<dd><p>The name of the grouping variable in meta.dat.</p></dd>


<dt id="arg-feature-level">feature.level<a class="anchor" aria-label="anchor" href="#arg-feature-level"></a></dt>
<dd><p>A character vector specifying taxonomic levels to be analyzed.</p></dd>


<dt id="arg-feature-mt-method">feature.mt.method<a class="anchor" aria-label="anchor" href="#arg-feature-mt-method"></a></dt>
<dd><p>Multiple testing method for features, "none" (default), "fdr", or other methods supported by p.adjust.</p></dd>


<dt id="arg-cutoff">cutoff<a class="anchor" aria-label="anchor" href="#arg-cutoff"></a></dt>
<dd><p>The p-value cutoff for significance.</p></dd>


<dt id="arg-color-group-level">color.group.level<a class="anchor" aria-label="anchor" href="#arg-color-group-level"></a></dt>
<dd><p>The taxonomic level used to color-code the branches of the cladogram.</p></dd>


<dt id="arg-palette">palette<a class="anchor" aria-label="anchor" href="#arg-palette"></a></dt>
<dd><p>An optional vector of colors to be used for the plot. If NULL, a default color palette will be used.</p></dd>


<dt id="arg-pdf">pdf<a class="anchor" aria-label="anchor" href="#arg-pdf"></a></dt>
<dd><p>Boolean indicating whether to save the plot as a PDF.</p></dd>


<dt id="arg-pdf-width">pdf.width<a class="anchor" aria-label="anchor" href="#arg-pdf-width"></a></dt>
<dd><p>The width of the PDF file if saved.</p></dd>


<dt id="arg-pdf-height">pdf.height<a class="anchor" aria-label="anchor" href="#arg-pdf-height"></a></dt>
<dd><p>The height of the PDF file if saved.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A ggplot object representing the circular heatmap with phylogenetic tree.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">subset_T2D.obj</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">test.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="generate_taxa_test_single.html">generate_taxa_test_single</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    data.obj <span class="op">=</span> <span class="va">subset_T2D.obj</span>,</span></span>
<span class="r-in"><span>    time.var <span class="op">=</span> <span class="st">"visit_number"</span>,</span></span>
<span class="r-in"><span>    t.level <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>    group.var <span class="op">=</span> <span class="st">"subject_race"</span>,</span></span>
<span class="r-in"><span>    adj.vars <span class="op">=</span> <span class="st">"subject_gender"</span>,</span></span>
<span class="r-in"><span>    feature.level <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Phylum"</span>, <span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"Family"</span>, <span class="st">"Genus"</span>, <span class="st">"Species"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    feature.dat.type <span class="op">=</span> <span class="st">"count"</span>,</span></span>
<span class="r-in"><span>    prev.filter <span class="op">=</span> <span class="fl">0.1</span>,</span></span>
<span class="r-in"><span>    abund.filter <span class="op">=</span> <span class="fl">0.0001</span>,</span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">plot.list</span> <span class="op">&lt;-</span> <span class="fu">generate_taxa_cladogram_single</span><span class="op">(</span></span></span>
<span class="r-in"><span>  data.obj <span class="op">=</span> <span class="va">subset_T2D.obj</span>,</span></span>
<span class="r-in"><span>  test.list <span class="op">=</span> <span class="va">test.list</span>,</span></span>
<span class="r-in"><span>  group.var <span class="op">=</span> <span class="st">"subject_gender"</span>,</span></span>
<span class="r-in"><span>  feature.level <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Phylum"</span>, <span class="st">"Class"</span>, <span class="st">"Order"</span>, <span class="st">"Family"</span>, <span class="st">"Genus"</span>, <span class="st">"Species"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  feature.mt.method <span class="op">=</span> <span class="st">"none"</span>,</span></span>
<span class="r-in"><span>  cutoff <span class="op">=</span> <span class="fl">0.9</span>,</span></span>
<span class="r-in"><span>  color.group.level <span class="op">=</span> <span class="st">"Order"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">test.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="generate_taxa_test_single.html">generate_taxa_test_single</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    data.obj <span class="op">=</span> <span class="va">subset_T2D.obj</span>,</span></span>
<span class="r-in"><span>    time.var <span class="op">=</span> <span class="st">"visit_number"</span>,</span></span>
<span class="r-in"><span>    t.level <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>    group.var <span class="op">=</span> <span class="st">"subject_race"</span>,</span></span>
<span class="r-in"><span>    adj.vars <span class="op">=</span> <span class="st">"subject_gender"</span>,</span></span>
<span class="r-in"><span>    feature.level <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Order"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    feature.dat.type <span class="op">=</span> <span class="st">"count"</span>,</span></span>
<span class="r-in"><span>    prev.filter <span class="op">=</span> <span class="fl">0.1</span>,</span></span>
<span class="r-in"><span>    abund.filter <span class="op">=</span> <span class="fl">0.0001</span>,</span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">plot.list</span> <span class="op">&lt;-</span> <span class="fu">generate_taxa_cladogram_single</span><span class="op">(</span></span></span>
<span class="r-in"><span>  data.obj <span class="op">=</span> <span class="va">subset_T2D.obj</span>,</span></span>
<span class="r-in"><span>  test.list <span class="op">=</span> <span class="va">test.list</span>,</span></span>
<span class="r-in"><span>  group.var <span class="op">=</span> <span class="st">"subject_gender"</span>,</span></span>
<span class="r-in"><span>  feature.level <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Order"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  feature.mt.method <span class="op">=</span> <span class="st">"none"</span>,</span></span>
<span class="r-in"><span>  cutoff <span class="op">=</span> <span class="fl">0.9</span>,</span></span>
<span class="r-in"><span>  color.group.level <span class="op">=</span> <span class="st">"Order"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">peerj32.obj</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">test.list</span> <span class="op">&lt;-</span> <span class="fu"><a href="generate_taxa_test_single.html">generate_taxa_test_single</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    data.obj <span class="op">=</span> <span class="va">peerj32.obj</span>,</span></span>
<span class="r-in"><span>    time.var <span class="op">=</span> <span class="st">"time"</span>,</span></span>
<span class="r-in"><span>    t.level <span class="op">=</span> <span class="cn">NULL</span>,</span></span>
<span class="r-in"><span>    group.var <span class="op">=</span> <span class="st">"group"</span>,</span></span>
<span class="r-in"><span>    adj.vars <span class="op">=</span> <span class="st">"sex"</span>,</span></span>
<span class="r-in"><span>    feature.level <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Phylum"</span>,<span class="st">"Family"</span>,<span class="st">"Genus"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    feature.dat.type <span class="op">=</span> <span class="st">"count"</span>,</span></span>
<span class="r-in"><span>    prev.filter <span class="op">=</span> <span class="fl">0.1</span>,</span></span>
<span class="r-in"><span>    abund.filter <span class="op">=</span> <span class="fl">0.0001</span>,</span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">plot.list</span> <span class="op">&lt;-</span> <span class="fu">generate_taxa_cladogram_single</span><span class="op">(</span></span></span>
<span class="r-in"><span>  data.obj <span class="op">=</span> <span class="va">peerj32.obj</span>,</span></span>
<span class="r-in"><span>  test.list <span class="op">=</span> <span class="va">test.list</span>,</span></span>
<span class="r-in"><span>  group.var <span class="op">=</span> <span class="st">"group"</span>,</span></span>
<span class="r-in"><span>  feature.level <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Phylum"</span>, <span class="st">"Family"</span>, <span class="st">"Genus"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  cutoff <span class="op">=</span> <span class="fl">0.3</span>,</span></span>
<span class="r-in"><span>  color.group.level <span class="op">=</span> <span class="st">"Family"</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Xianyang Zhang, Jun Chen, Caffery(Chen) Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

