<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate various beta diversity indices — mStat_calculate_beta_diversity • MicrobiomeStat</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate various beta diversity indices — mStat_calculate_beta_diversity"><meta name="description" content="This function calculates a variety of beta diversity indices based on the input data, such as
Bray-Curtis (BC), Jaccard, unweighted UniFrac (UniFrac), generalized UniFrac (GUniFrac),
weighted UniFrac (WUniFrac), and Jensen-Shannon divergence (JS)."><meta property="og:description" content="This function calculates a variety of beta diversity indices based on the input data, such as
Bray-Curtis (BC), Jaccard, unweighted UniFrac (UniFrac), generalized UniFrac (GUniFrac),
weighted UniFrac (WUniFrac), and Jensen-Shannon divergence (JS)."><meta property="og:image" content="https://cafferychen777.github.io/MicrobiomeStat/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MicrobiomeStat</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/cafferychen777/MicrobiomeStat"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Calculate various beta diversity indices</h1>
      <small class="dont-index">Source: <a href="https://github.com/cafferychen777/MicrobiomeStat/blob/main/R/mStat_calculate_beta_diversity.R" class="external-link"><code>R/mStat_calculate_beta_diversity.R</code></a></small>
      <div class="d-none name"><code>mStat_calculate_beta_diversity.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function calculates a variety of beta diversity indices based on the input data, such as
Bray-Curtis (BC), Jaccard, unweighted UniFrac (UniFrac), generalized UniFrac (GUniFrac),
weighted UniFrac (WUniFrac), and Jensen-Shannon divergence (JS).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">mStat_calculate_beta_diversity</span><span class="op">(</span></span>
<span>  <span class="va">data.obj</span>,</span>
<span>  dist.name <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"BC"</span>, <span class="st">"Jaccard"</span>, <span class="st">"UniFrac"</span>, <span class="st">"GUniFrac"</span>, <span class="st">"WUniFrac"</span>, <span class="st">"JS"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data-obj">data.obj<a class="anchor" aria-label="anchor" href="#arg-data-obj"></a></dt>
<dd><p>A list object in a format specific to MicrobiomeStat, which can include components such as feature.tab (matrix), feature.ann (matrix), meta.dat (data.frame), tree, and feature.agg.list (list). The data.obj can be converted from other formats using several functions from the MicrobiomeStat package, including: 'mStat_convert_DGEList_to_data_obj', 'mStat_convert_DESeqDataSet_to_data_obj', 'mStat_convert_phyloseq_to_data_obj', 'mStat_convert_SummarizedExperiment_to_data_obj', 'mStat_import_qiime2_as_data_obj', 'mStat_import_mothur_as_data_obj', 'mStat_import_dada2_as_data_obj', and 'mStat_import_biom_as_data_obj'. Alternatively, users can construct their own data.obj. Note that not all components of data.obj may be required for all functions in the MicrobiomeStat package.</p></dd>


<dt id="arg-dist-name">dist.name<a class="anchor" aria-label="anchor" href="#arg-dist-name"></a></dt>
<dd><p>A character vector specifying which beta diversity indices to calculate. Supported indices are "BC" (Bray-Curtis), "Jaccard", "UniFrac" (unweighted UniFrac), "GUniFrac" (generalized UniFrac), "WUniFrac" (weighted UniFrac), and "JS" (Jensen-Shannon divergence). If a name is provided but the corresponding object does not exist within dist.obj, it will be computed internally. If the specific index is not supported, an error message will be returned.
"BC" (Bray-Curtis), "Jaccard", "UniFrac" (unweighted UniFrac), "GUniFrac" (generalized UniFrac),
"WUniFrac" (weighted UniFrac), and "JS" (Jensen-Shannon divergence).</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing the calculated beta diversity indices. The indices are named with the abbreviation.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/vegandevs/vegan" class="external-link">vegan</a></span><span class="op">)</span> <span class="co"># Used for community ecology analyses</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GUniFrac</span><span class="op">)</span> <span class="co"># For generalized UniFrac distances</span></span></span>
<span class="r-in"><span><span class="co"># library(phyloseq) # For handling and analyzing phylogenetic sequencing data</span></span></span>
<span class="r-in"><span><span class="co"># Load example data</span></span></span>
<span class="r-in"><span><span class="co"># data(GlobalPatterns) # An example dataset from the phyloseq package</span></span></span>
<span class="r-in"><span><span class="co"># Convert the phyloseq object to a MicrobiomeStat data object</span></span></span>
<span class="r-in"><span><span class="co"># This step is crucial for making the dataset compatible with MicrobiomeStat functions</span></span></span>
<span class="r-in"><span><span class="co"># GlobalPatterns.obj &lt;- mStat_convert_phyloseq_to_data_obj(GlobalPatterns)</span></span></span>
<span class="r-in"><span><span class="co"># Calculate various beta diversity indices</span></span></span>
<span class="r-in"><span><span class="co"># Beta diversity measures the difference in microbial communities across samples</span></span></span>
<span class="r-in"><span><span class="co"># Bray-Curtis dissimilarity (BC)</span></span></span>
<span class="r-in"><span><span class="co"># A commonly used measure of dissimilarity based on counts</span></span></span>
<span class="r-in"><span><span class="co"># dist.obj &lt;- mStat_calculate_beta_diversity(GlobalPatterns.obj, dist.name = c('BC'))</span></span></span>
<span class="r-in"><span><span class="co"># Jaccard index</span></span></span>
<span class="r-in"><span><span class="co"># A measure based on presence/absence, useful for binary data</span></span></span>
<span class="r-in"><span><span class="co"># dist.obj &lt;- mStat_calculate_beta_diversity(GlobalPatterns.obj, dist.name = c("Jaccard"))</span></span></span>
<span class="r-in"><span><span class="co"># UniFrac distance</span></span></span>
<span class="r-in"><span><span class="co"># A phylogenetic measure of community dissimilarity</span></span></span>
<span class="r-in"><span><span class="co"># Requires a phylogenetic tree as part of the input</span></span></span>
<span class="r-in"><span><span class="co"># dist.obj &lt;- mStat_calculate_beta_diversity(GlobalPatterns.obj, dist.name = c('UniFrac'))</span></span></span>
<span class="r-in"><span><span class="co"># Weighted UniFrac distance (WUniFrac)</span></span></span>
<span class="r-in"><span><span class="co"># A variation of UniFrac that accounts for relative abundance</span></span></span>
<span class="r-in"><span><span class="co"># Also requires a phylogenetic tree</span></span></span>
<span class="r-in"><span><span class="co"># dist.obj &lt;- mStat_calculate_beta_diversity(GlobalPatterns.obj, dist.name = c('WUniFrac'))</span></span></span>
<span class="r-in"><span><span class="co"># Jensen-Shannon divergence (JS)</span></span></span>
<span class="r-in"><span><span class="co"># A symmetric and smoothed version of the Kullback-Leibler divergence</span></span></span>
<span class="r-in"><span><span class="co"># Useful for comparing probability distributions</span></span></span>
<span class="r-in"><span><span class="co"># dist.obj &lt;- mStat_calculate_beta_diversity(GlobalPatterns.obj, dist.name = c('JS'))</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Xianyang Zhang, Jun Chen, Caffery(Chen) Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

