<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Normalize a MicrobiomeStat Data Object — mStat_normalize_data • MicrobiomeStat</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Normalize a MicrobiomeStat Data Object — mStat_normalize_data"><meta name="description" content="This function is part of the MicrobiomeStat package. It normalizes a data object based on the chosen method."><meta property="og:description" content="This function is part of the MicrobiomeStat package. It normalizes a data object based on the chosen method."><meta property="og:image" content="https://cafferychen777.github.io/MicrobiomeStat/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="default" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">MicrobiomeStat</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.2.1</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/cafferychen777/MicrobiomeStat"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Normalize a MicrobiomeStat Data Object</h1>
      <small class="dont-index">Source: <a href="https://github.com/cafferychen777/MicrobiomeStat/blob/main/R/mStat_normalize_data.R" class="external-link"><code>R/mStat_normalize_data.R</code></a></small>
      <div class="d-none name"><code>mStat_normalize_data.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function is part of the MicrobiomeStat package. It normalizes a data object based on the chosen method.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">mStat_normalize_data</span><span class="op">(</span></span>
<span>  <span class="va">data.obj</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Rarefy-TSS"</span>, <span class="st">"Rarefy"</span>, <span class="st">"TSS"</span>, <span class="st">"GMPR"</span>, <span class="st">"CSS"</span>, <span class="st">"DESeq"</span>, <span class="st">"TMM"</span><span class="op">)</span>,</span>
<span>  depth <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-data-obj">data.obj<a class="anchor" aria-label="anchor" href="#arg-data-obj"></a></dt>
<dd><p>A list object in a format specific to MicrobiomeStat, which can include components such as feature.tab (matrix), feature.ann (matrix), meta.dat (data.frame), tree, and feature.agg.list (list). The data.obj can be converted from other formats using several functions from the MicrobiomeStat package, including: 'mStat_convert_DGEList_to_data_obj', 'mStat_convert_DESeqDataSet_to_data_obj', 'mStat_convert_phyloseq_to_data_obj', 'mStat_convert_SummarizedExperiment_to_data_obj', 'mStat_import_qiime2_as_data_obj', 'mStat_import_mothur_as_data_obj', 'mStat_import_dada2_as_data_obj', and 'mStat_import_biom_as_data_obj'. Alternatively, users can construct their own data.obj. Note that not all components of data.obj may be required for all functions in the MicrobiomeStat package.</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>A string. The normalization method to be applied. It must be one of the following: "Rarefy-TSS", "Rarefy", "TSS", "GMPR", "CSS", "DESeq", "TMM". The default is "Rarefy-TSS".
- "Rarefy-TSS": Rarefaction followed by Total Sum Scaling normalization.
- "Rarefy": Rarefaction normalization only.
- "TSS": Total Sum Scaling normalization only.
- "GMPR": Geometric Mean of Pairwise Ratios normalization method.
- "CSS": Cumulative Sum Scaling normalization method.
- "DESeq": Normalization using the DESeq method for RNA-seq data.
- "TMM": Normalization using the Trimmed Mean of M-values (TMM) method from the edgeR package.</p></dd>


<dt id="arg-depth">depth<a class="anchor" aria-label="anchor" href="#arg-depth"></a></dt>
<dd><p>An integer. The sequencing depth to be used for the "Rarefy" and "Rarefy-TSS" methods. If NULL, the smallest total count dplyr::across samples is used as the rarefaction depth.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list. The normalized data object and the scale factor used for normalization.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The function first checks if 'data.obj' is a list. It then retrieves the OTU table and estimates the normalization/scale factor based on the chosen method. The data object is then updated with the normalized OTU table and the chosen method is added as 'norm.status'. The function returns the normalized data object and the scale factor.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span> <span class="co"># \dontrun{</span></span></span>
<span class="r-in"><span><span class="co"># Load example data object</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">peerj32.obj</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Applying Total Sum Scaling (TSS) normalization</span></span></span>
<span class="r-in"><span><span class="va">norm_result_tss</span> <span class="op">&lt;-</span> <span class="fu">mStat_normalize_data</span><span class="op">(</span>data.obj <span class="op">=</span> <span class="va">peerj32.obj</span>, method <span class="op">=</span> <span class="st">"TSS"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">norm_result_tss</span><span class="op">$</span><span class="va">data.obj.norm</span><span class="op">)</span>  <span class="co"># Display normalized data object</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Applying Rarefaction followed by Total Sum Scaling (Rarefy-TSS) with a specified depth</span></span></span>
<span class="r-in"><span><span class="va">norm_result_rarefy_tss</span> <span class="op">&lt;-</span> <span class="fu">mStat_normalize_data</span><span class="op">(</span>data.obj <span class="op">=</span> <span class="va">peerj32.obj</span>,</span></span>
<span class="r-in"><span>method <span class="op">=</span> <span class="st">"Rarefy-TSS"</span>, depth <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">norm_result_rarefy_tss</span><span class="op">$</span><span class="va">data.obj.norm</span><span class="op">)</span>  <span class="co"># Display normalized data object</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Normalization using Geometric Mean of Pairwise Ratios (GMPR)</span></span></span>
<span class="r-in"><span><span class="va">norm_result_gmpr</span> <span class="op">&lt;-</span> <span class="fu">mStat_normalize_data</span><span class="op">(</span>data.obj <span class="op">=</span> <span class="va">peerj32.obj</span>, method <span class="op">=</span> <span class="st">"GMPR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">norm_result_gmpr</span><span class="op">$</span><span class="va">data.obj.norm</span><span class="op">)</span>  <span class="co"># Display normalized data object</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Utilizing the DESeq normalization method</span></span></span>
<span class="r-in"><span><span class="co"># This is particularly useful for RNA-seq data from microbiome studies</span></span></span>
<span class="r-in"><span><span class="va">norm_result_deseq</span> <span class="op">&lt;-</span> <span class="fu">mStat_normalize_data</span><span class="op">(</span>data.obj <span class="op">=</span> <span class="va">peerj32.obj</span>, method <span class="op">=</span> <span class="st">"DESeq"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">norm_result_deseq</span><span class="op">$</span><span class="va">data.obj.norm</span><span class="op">)</span>  <span class="co"># Display normalized data object</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Example of error handling when an incorrect depth is specified for the "Rarefy" method</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/conditions.html" class="external-link">tryCatch</a></span><span class="op">(</span><span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="va">norm_result_error</span> <span class="op">&lt;-</span> <span class="fu">mStat_normalize_data</span><span class="op">(</span>data.obj <span class="op">=</span> <span class="va">peerj32.obj</span>,</span></span>
<span class="r-in"><span>  method <span class="op">=</span> <span class="st">"Rarefy"</span>, depth <span class="op">=</span> <span class="fl">10000000</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">norm_result_error</span><span class="op">$</span><span class="va">data.obj.norm</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span>, error <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">e</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">e</span><span class="op">$</span><span class="va">message</span><span class="op">)</span>  <span class="co"># Print the error message if depth is not feasible</span></span></span>
<span class="r-in"><span><span class="op">}</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="co"># }</span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Xianyang Zhang, Jun Chen, Caffery(Chen) Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

