<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="MicrobiomeStat is a comprehensive R package for microbiome analysis, designed specifically for longitudinal studies. It provides solutions for longitudinal, paired, and cross-sectional analyses, and is capable of handling multi-omics data. The package aims to facilitate efficient and effective analysis of multi-omics data in longitudinal studies, making it suitable for a wide range of applications. MicrobiomeStat is designed with user-friendliness in mind, offering a range of features and practical use cases to support researchers in their work. Stay up-to-date with the latest updates and experience the benefits of using MicrobiomeStat for your microbiome analysis needs.">
<title>Comprehensive Statistical and Visualization Methods for Microbiome and Multi-Omics Data â€¢ MicrobiomeStat</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Comprehensive Statistical and Visualization Methods for Microbiome and Multi-Omics Data">
<meta property="og:description" content="MicrobiomeStat is a comprehensive R package for microbiome analysis, designed specifically for longitudinal studies. It provides solutions for longitudinal, paired, and cross-sectional analyses, and is capable of handling multi-omics data. The package aims to facilitate efficient and effective analysis of multi-omics data in longitudinal studies, making it suitable for a wide range of applications. MicrobiomeStat is designed with user-friendliness in mind, offering a range of features and practical use cases to support researchers in their work. Stay up-to-date with the latest updates and experience the benefits of using MicrobiomeStat for your microbiome analysis needs.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">MicrobiomeStat</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/cafferychen777/MicrobiomeStat/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="microbiomestat-your-best-choice-for-microbiome-analysis-ever-">MicrobiomeStat: Your Best Choice for Microbiome Analysis Ever! ğŸ§¬ğŸ”¬<a class="anchor" aria-label="anchor" href="#microbiomestat-your-best-choice-for-microbiome-analysis-ever-"></a>
</h1></div>
<!-- badges: start -->

<p><code>MicrobiomeStat</code> is a powerful and versatile R package, designed specifically for analyzing longitudinal microbiome data. It extends to handle multi-omics data, making it an asset for broader biological research.</p>
<p>Crucially, <code>MicrobiomeStat</code> shines not only in longitudinal studies, but is also adept at handling cross-sectional studies. For more detailed tutorials and information, please visit our main page: <a href="http://www.microbiomestat.wiki" class="external-link">www.microbiomestat.wiki</a>.</p>
<div class="section level2">
<h2 id="why-choose-microbiomestat-">Why Choose MicrobiomeStat? ğŸ‘<a class="anchor" aria-label="anchor" href="#why-choose-microbiomestat-"></a>
</h2>
<p>Choosing the right tools for your microbiome data analysis can make a significant difference in the results. We have compared <code>MicrobiomeStat</code> with other similar packages, and we believe our package stands out in many aspects. Below are two comparative tables, showing the distinct advantages of <code>MicrobiomeStat</code> over other packages.</p>
</div>
<div class="section level2">
<h2 id="comparative-table-1-longitudinal-packages">Comparative Table 1: Longitudinal Packages<a class="anchor" aria-label="anchor" href="#comparative-table-1-longitudinal-packages"></a>
</h2>
<table class="table">
<colgroup>
<col width="14%">
<col width="27%">
<col width="13%">
<col width="22%">
<col width="22%">
</colgroup>
<thead><tr class="header">
<th align="center">Features</th>
<th align="center">MicrobiomeStat</th>
<th align="center">q2-longitudinal</th>
<th align="center">SplinectomeR</th>
<th align="center">coda4microbiome</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">Data Input Formats</td>
<td align="center">QIIME2, DADA2, BIOM, Mothur, Phyloseq, DGEList, DESeqDataSet, SummarizedExperiment, etc.</td>
<td align="center">QIIME2</td>
<td align="center">Custom format (No other formats import or conversion functions provided)</td>
<td align="center">Custom format (No other formats import or conversion functions provided)</td>
</tr>
<tr class="even">
<td align="center">Supports Alpha Diversity Analysis</td>
<td align="center">âœ”ï¸</td>
<td align="center">Metrics need to be calculated externally</td>
<td align="center">Metrics need to be calculated externally</td>
<td align="center">âŒ</td>
</tr>
<tr class="odd">
<td align="center">Supports Beta Diversity Analysis</td>
<td align="center">âœ”ï¸</td>
<td align="center">Distances need to be calculated externally</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
</tr>
<tr class="even">
<td align="center">Paired Sample Analysis</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
</tr>
<tr class="odd">
<td align="center">Multiple Visualization Styles</td>
<td align="center">Boxplot, Violinplot, Spaghettiplot, Heatmap, Stacked Barplot, Areaplot, Volcano Plot</td>
<td align="center">Spaghettiplot, Boxplot</td>
<td align="center">Lineplot</td>
<td align="center">Boxplot, Densityplot, Barplot, Lineplot</td>
</tr>
<tr class="even">
<td align="center">Key Features Identification</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
</tr>
<tr class="odd">
<td align="center">Diff. Prevalence</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
</tr>
<tr class="even">
<td align="center">Paired Samples Support in Taxa Analysis</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
</tr>
<tr class="odd">
<td align="center">Time Points Subgroup Analysis in Taxa Analysis</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
</tr>
<tr class="even">
<td align="center">Similar Abundance Pattern</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
</tr>
<tr class="odd">
<td align="center">Similar Change Pattern</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
</tr>
<tr class="even">
<td align="center">Similar Changes in Composition Over Time</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
</tr>
<tr class="odd">
<td align="center">Publication Ready</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
</tr>
<tr class="even">
<td align="center">Ongoing Development</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
</tr>
</tbody>
</table>
<p>As shown in Comparative Table 1, <code>MicrobiomeStat</code> provides a variety of data input formats, supports both alpha and beta diversity analysis, paired sample analysis, multiple visualization styles, and many other features. This makes <code>MicrobiomeStat</code> more comprehensive and flexible than many other packages dedicated to longitudinal data analysis.</p>
</div>
<div class="section level2">
<h2 id="comparative-table-2-comprehensive-packages">Comparative Table 2: Comprehensive Packages<a class="anchor" aria-label="anchor" href="#comparative-table-2-comprehensive-packages"></a>
</h2>
<table style="width:100%;" class="table">
<colgroup>
<col width="4%">
<col width="13%">
<col width="9%">
<col width="6%">
<col width="5%">
<col width="16%">
<col width="11%">
<col width="13%">
<col width="2%">
<col width="4%">
<col width="12%">
</colgroup>
<thead><tr class="header">
<th align="center">Features</th>
<th align="center">MicrobiomeStat</th>
<th align="center">microbiomeutilities</th>
<th align="center">phyloseq</th>
<th align="center">microbiomemarker</th>
<th align="center">MicrobiomeAnalyst</th>
<th align="center">microbiomeeco</th>
<th align="center">EasyAmplicon</th>
<th align="center">STAMP</th>
<th align="center">qiime2</th>
<th align="center">MicrobiotaProcess</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">Data Input Formats</td>
<td align="center">QIIME2, DADA2, BIOM, Mothur, Phyloseq, DGEList, DESeqDataSet, SummarizedExperiment</td>
<td align="center">phyloseq</td>
<td align="center">QIIME2, DADA2, BIOM, Mothur</td>
<td align="center">QIIME2, DADA2, BIOM, Mothur, Phyloseq</td>
<td align="center">txt, .csv, biom, mothur</td>
<td align="center">QIIME, QIIME2, HUMAnN, Kraken2, phyloseq</td>
<td align="center">Paired fastq/fq files</td>
<td align="center">.tsv</td>
<td align="center">fastq, biom</td>
<td align="center">qiime2, dada2, MetaPhlAn, Returns MPSE object, Phyloseq, TreeSummarizedExperiment, SummarizedExperiment</td>
</tr>
<tr class="even">
<td align="center">Supports Alpha Diversity Analysis</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
</tr>
<tr class="odd">
<td align="center">Supports Beta Diversity Analysis</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
</tr>
<tr class="even">
<td align="center">Paired Sample Analysis</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
</tr>
<tr class="odd">
<td align="center">Longitudinal Data Analysis</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
</tr>
<tr class="even">
<td align="center">Multiple Visualization Styles</td>
<td align="center">Boxplot, Violinplot, Scatterplot, Barplot, Heatmap, Dotplot, Ordination plot, Area plot, Volcano plot, Spaghetti plot</td>
<td align="center">Density plot, Barplot, Stripchart, Boxplot, Heatmap, Area plot, Spaghetti plot</td>
<td align="center">Boxplot, Barplot, Heatmap, Network plot, Ordination plot</td>
<td align="center">Boxplot, Heatmap, Dotplot, Barplot, Cladogram</td>
<td align="center">Stacked bar/area plot, Interactive pie chart, Rarefaction curve, Phylogenetic tree, Heat tree, Boxplot, Heatmap, Network plot, Ordination plot</td>
<td align="center">Barplot, Boxplot, Heatmap, Pie chart, Venn diagram, Ordination plot, Circular heatmap, Sankey diagram</td>
<td align="center">Barplot, Cladogram, Phylogenetic tree, Heatmap, Sankey diagram, Venn diagram, Ordination plot, Volcano plot, Boxplot</td>
<td align="center">Barplot, Scatterplot</td>
<td align="center">Scatterplot, Ordination plot, Boxplot</td>
<td align="center">Barplot, Boxplot, Heatmap, Ordination plot, Tree plot, Volcano plot</td>
</tr>
<tr class="odd">
<td align="center">Publication Ready</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âŒ</td>
</tr>
<tr class="even">
<td align="center">Ongoing Development</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
<td align="center">âŒ</td>
<td align="center">âœ”ï¸</td>
<td align="center">âœ”ï¸</td>
</tr>
</tbody>
</table>
<p>Comparative Table 2, comparing comprehensive packages, further illustrates the strengths of <code>MicrobiomeStat</code>. Not only does <code>MicrobiomeStat</code> support a wide range of data input formats and diversity analyses, but it also excels in paired sample analysis and longitudinal data analysis, among others.</p>
<p>In conclusion, <code>MicrobiomeStat</code> offers a robust and comprehensive solution for your microbiome data analysis needs, outperforming many other packages in both versatility and user-friendliness. Why not give it a try and see the difference for yourself?</p>
</div>
<div class="section level2">
<h2 id="features-at-a-glance-">Features at a glance ğŸŒŸ<a class="anchor" aria-label="anchor" href="#features-at-a-glance-"></a>
</h2>
<table class="table">
<colgroup>
<col width="24%">
<col width="75%">
</colgroup>
<thead><tr class="header">
<th>Feature</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Data Import and Conversion</td>
<td>Supports numerous input formats from popular tools like QIIME2, Mothur, DADA2, Phyloseq and more</td>
</tr>
<tr class="even">
<td>Cross-sectional Study Analysis</td>
<td>Performs comprehensive analysis of cross-sectional studies</td>
</tr>
<tr class="odd">
<td>Paired Sample Analysis</td>
<td>Excellent tool for analyzing paired samples</td>
</tr>
<tr class="even">
<td>Longitudinal Study Analysis</td>
<td>Allows for exploring the temporal dynamics of the microbiome</td>
</tr>
<tr class="odd">
<td>Visualization Capabilities</td>
<td>Offers a wide variety of visualization styles</td>
</tr>
<tr class="even">
<td>Data Export</td>
<td>Supports export of analysis results in diverse formats</td>
</tr>
<tr class="odd">
<td>Ongoing Development</td>
<td>Continual feature refinement and new functionality addition</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="output-interpretation-">Output interpretation ğŸ“<a class="anchor" aria-label="anchor" href="#output-interpretation-"></a>
</h2>
<p>To learn how to interpret the output objects and visualizations, visit our wiki: <a href="http://www.microbiomestat.wiki" class="external-link">www.microbiomestat.wiki</a>.</p>
</div>
<div class="section level2">
<h2 id="known-issues-and-limitations-ï¸">Known issues and limitations â—ï¸<a class="anchor" aria-label="anchor" href="#known-issues-and-limitations-%EF%B8%8F"></a>
</h2>
<p>We strive to make <code>MicrobiomeStat</code> as comprehensive as possible, but there may still be limitations. All known issues can be found in the issues section of our GitHub repository.</p>
</div>
<div class="section level2">
<h2 id="support--contact-">Support &amp; Contact ğŸ“¬<a class="anchor" aria-label="anchor" href="#support--contact-"></a>
</h2>
<table class="table">
<thead><tr class="header">
<th>Name</th>
<th>Email</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Jun Chen</td>
<td><a href="mailto:Chen.Jun2@mayo.edu" class="email">Chen.Jun2@mayo.edu</a></td>
</tr>
<tr class="even">
<td>Caffery Yang</td>
<td><a href="mailto:cafferychen7850@gmail.com" class="email">cafferychen7850@gmail.com</a></td>
</tr>
</tbody>
</table>
<p>Stay updated with the latest <code>MicrobiomeStat</code> features by visiting our wiki: <a href="http://www.microbiomestat.wiki" class="external-link">www.microbiomestat.wiki</a></p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=MicrobiomeStat" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/cafferychen777/MicrobiomeStat/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/cafferychen777/MicrobiomeStat/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing MicrobiomeStat</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Xianyang Zhang <br><small class="roles"> Author </small>  </li>
<li>Jun Chen <br><small class="roles"> Author, maintainer </small>  </li>
<li>Caffery(Chen) Yang <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://cran.r-project.org/package=MicrobiomeStat" class="external-link"><img src="https://www.r-pkg.org/badges/version/MicrobiomeStat" alt="CRAN_Status_Badge"></a></li>
<li><a href="https://cran.r-project.org/package=MicrobiomeStat" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/grand-total/MicrobiomeStat" alt="Downloads"></a></li>
<li><a href="https://www.gnu.org/licenses/gpl-3.0" class="external-link"><img src="https://img.shields.io/badge/License-GPLv3-blue.svg" alt="License: GPL v3"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Xianyang Zhang, Jun Chen, Caffery(Chen) Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
